api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -242,9 +249,9 @@ public static Cohort intersect(Cohort a, Cohort b) {
 	public static Cohort subtract(Cohort a, Cohort b) {
 		Cohort ret = new Cohort();
 		if (a != null) {
-			ret.getMemberIds().addAll(a.getMemberIds());
+			ret.getMembers().addAll(a.getMembers());
 			if (b != null) {
-				ret.getMemberIds().removeAll(b.getMemberIds());
+                ret.getMembers().removeAll(b.getMembers());
```

Comment:
The issue I'm pointing out here is to do with equality and I'm thinking about it from the perspective of any Collection in general, collections use the equals method to find the item to be removed. The equals and hashcode() methods from the super class i.e BaseOpenmrsObject are based on uuid, when you call removeAll(), meaning the member to be removed will be one with the same uuid but what you really want to remove are members referencing the same Patient.